### DISCUSSION TITLE
Improve TileJSON response with CDN and 2x tiles

### DISCUSSION THREAD

**Discussion by maxmalynowsky** [developmentseed/titiler#621] [Q&A]
## Problem description

👋 Hey all!

I have TiTiler deployed on an Azure Container App behind a CDN, serving as the backend to a few MapLibre GL maps. The easiest way for me to add data to the maps would be to get a TileJSON, but there are two issues I encountered. This is an example URL I would try:

Request:

`https://titiler-cdn.mydomain.com/cog/tilejson.json?tile_format=webp&tile_scale=2&url=https://data.mydomain.com/data.tif`

Response:

```JSON
{
  "tilejson": "2.2.0",
  "version": "1.0.0",
  "scheme": "xyz",
  "tiles": [
    "http://myapp.azurecontainerapps.io/cog/tiles/WebMercatorQuad/{z}/{x}/{y}@2x.webp?url=https%3A%2F%2Fmydomain.com%2Fdata.tif"
  ],
  "minzoom": 8,
  "maxzoom": 15,
  "bounds": [-1, -1, 1, 1],
  "center": [0, 0, 8]
}
```

Issues:

1. The URL in `tiles` property comes from the URL my container is deployed to in Azure, but I'd like it to use the request origin of the CDN instead. I'm a bit new at developing with Starlette / FastAPI / TiTiler, so I'm not sure if there's a way to set this as an ENV property anywhere, which would be ideal. Otherwise, I wouldn't mind being able to set a `tile_url` property such as `https://titiler-cdn.mydomain.com` in the query to specify this.
2. In this example, my GeoTIFF can be tiled down to zoom 15 with 256px tiles. However, using 512px tiles, it should have a max zoom of 14. I made an [edit](https://github.com/maxmalynowsky/titiler/blob/4b1f69e30b0017b76f92512526d86acb29a35875/src/titiler/core/titiler/core/factory.py#L602) suggesting the way this could be implemented, but I'm still figuring out how to set this up locally to test.

**Comment by vincentsarago** [comment]
@maxmalynowsky can you check https://github.com/developmentseed/titiler/discussions/342 and see if you find useful information

**Comment by maxmalynowsky** [comment]
Hi @vincentsarago , thanks for the link, but setting `root_path` doesn't address the problem I'm having. 

I'm not looking to change `https://mydomain.com/cog/tilejson.json` to `https://mydomain.com/api/v1/cog/tilejson.json`
but rather to change `http://azureapp.io/cog/tilejson.json` to `https://mydomain.com/cog/tilejson.json`

After doing more research into this, it seems my issue comes from Starlette's `request.base_url` as used [here](https://github.com/developmentseed/titiler/blob/649788a0fb3acbccea4ebc9527a318c77fb0b1c7/src/titiler/core/titiler/core/factory.py#L1375) in making an absolute URL. I'm guessing the solution would need to come from implementing custom middleware to set `request._base_url` to the URL I want to have included in the tilejson response. I realize this is unrelated to titiler and not something that's likely easy to implement, although if you've seen a potential fix for this it would really help.

**Comment by vincentsarago** [comment]
there is also this discussion https://github.com/developmentseed/titiler/discussions/345#discussioncomment-1636050 which could be useful

**Comment by maxmalynowsky** [comment]
I saw that issue, unfortunately `--forwarded-allow-ips=*` didn't result in a change. Worryingly, setting that also failed to change the scheme of my request from `http` to `https`. This issue seems like it is the closest to solving what I needed, but implementing it had no effect: https://stackoverflow.com/questions/64600957/fastapi-set-the-url-or-path-for-the-request-object

**Comment by vincentsarago** [comment]
Not sure if it helps, but in AWS I remember I had to do this 👇 
![Screen Shot 2021-03-09 at 10 46 58 PM](https://user-images.githubusercontent.com/10407788/204301758-8ce775e7-87ba-4b0f-8c91-bc6d65ec5fe3.png)
